ir/base.def

```diff
@@ -332,17 +332,36 @@ class Annotation {
             Vector<Expression>,
             IndexedVector<NamedExpression>> body;
 
-    inline auto &getUnparsed() { return std::get<UnparsedAnnotation>(body); }
-    inline const auto &getUnparsed() const { return std::get<UnparsedAnnotation>(body); }
-    inline auto &getExpr() { return std::get<ExpressionAnnotation>(body); }
-    inline const auto &getExpr() const { return std::get<ExpressionAnnotation>(body); }
+    inline auto &getUnparsed() {
+        BUG_CHECK(annotationKind() == Kind::Unparsed, "Annotation has been parsed already.");
+        return std::get<UnparsedAnnotation>(body);
+    }
+    inline const auto &getUnparsed() const {
+        BUG_CHECK(annotationKind() == Kind::Unparsed, "Annotation has been parsed already.");
+        return std::get<UnparsedAnnotation>(body);
+    }
+    inline auto &getExpr() {
+        BUG_CHECK(annotationKind() == Kind::Unstructured || annotationKind() == Kind::StructuredExpressionList, "Annotation does not contain an expression list.");
+        return std::get<ExpressionAnnotation>(body);
+     }
+    inline const auto &getExpr() const {
+        BUG_CHECK(annotationKind() == Kind::Unstructured || annotationKind() == Kind::StructuredExpressionList, "Annotation does not contain an expression list.");
+        return std::get<ExpressionAnnotation>(body);
+    }
     inline Expression getExpr(size_t idx) const {
+        BUG_CHECK(annotationKind() == Kind::Unstructured || annotationKind() == Kind::StructuredExpressionList, "Annotation does not contain an expression list.");
         const auto &expr = getExpr();
         BUG_CHECK(idx < expr.size(), "invalid annotation expression index");
         return expr[idx];
     }
-    inline auto &getKV() { return std::get<KVAnnotation>(body); }
-    inline const auto &getKV() const { return std::get<KVAnnotation>(body); }
+    inline auto &getKV() {
+        BUG_CHECK(annotationKind() == Kind::StructuredKVList, "Annotation does not contain a key-value list.");
```

Comment:
What if the annotation is an unstructured kv-list (an unstructured annotation with `PARSE_KV_LIST`)

Reply:
Will add this. Are these all possible combinations?

Reply:
When an unstructured annotation is parsed, it might be parsed into either the `expr` field or the `kv` field, or into something else, though I don't think we've run into that yet.  The original intent was that a backend that needed such an annotation would define a subclass of `Annotation` to store the parsed version of the annotation.

Reply:
When do we generate a `structured` annotation? I have been looking at 
```
IR::Annotation::Kind annotationKind() const {
    if (needsParsing())
        return Kind::Unparsed;
    if (!structured)
        return Kind::Unstructured;
    if (std::holds_alternative<ExpressionAnnotation>(body))
        return Kind::StructuredExpressionList;
    if (std::holds_alternative<KVAnnotation>(body))
        return Kind::StructuredKVList;

    BUG("Invalid annotation kind");
}
```
and it is a little misleading because an annotation can be both a `Kind::StructuredKVList`/`Kind::StructuredExpressionList` and a `Kind::Unstructured`

Reply:
Looking at https://p4.org/p4-spec/docs/P4-16-working-spec.html#sec-annotations it is possible that a compiler developers tries to parse an unstructured annotation as a KV-annotation. We do not restrict that, right?

Reply:
Looks like with Anton's change, one needs to check `std::holds_alternative` to check that it has the right kind (or does using `std::get` throw an exception if the wrong kind is present?  That would probably be ok).

Reply:
The `structured` flag is set if the code uses `@foo [ ... ]` or `@foo { ... }` and not set if it uses `@foo( ... )`.  That's all it really tells you.

Reply:
> and it is a little misleading because an annotation can be both a

Yes, I was pretty confused by this distinction. Even more, it seems to be used just in one place – somewhere around p4 control plane. Overall, the things are multimensional:

 - Annotation could be parsed and unparsed
 - Parsed annotations could be structured and unstructured (1 dimension)
 - Parsed annotations could be expression lists and kv-lists (2 dimension)

I think it would be better to factor out everything in such way. I just do not know if there are any out of tree uses of structured annotations to get rid of `annotationKind`.